optional_plugins/golang/avocado_golang/__init__.py

```diff
@@ -0,0 +1,183 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2017
+# Authors: Amador Pahim <apahim@redhat.com>
+
+"""
+Plugin to run Golang tests in Avocado
+"""
+
+import fnmatch
+import os
+import re
+
+from avocado.core import loader
+from avocado.core import output
+from avocado.core import test
+from avocado.core.plugin_interfaces import CLI
+from avocado.utils import path as utils_path
+from avocado.utils import process
+
+
+class GolangTest(test.SimpleTest):
+
+    """
+    Run a Golang Test command as a SIMPLE test.
+    """
+
+    def __init__(self, name,
+                 params=None,
+                 base_logdir=None,
+                 job=None,
+                 subtest=None):
+        super(GolangTest, self).__init__(name, params, base_logdir, job)
+        self.subtest = subtest
+
+    @property
+    def filename(self):
+        """
+        Returns the path of the golang test suite.
+        """
+        return self.name.name.split(':')[0]
+
+    def test(self):
+        """
+        Create the Golang command and execute it.
+        """
+        test_name = '%s$' % self.name.name.split(':')[1]
+        if self.subtest is not None:
+            test_name += '/%s' % self.subtest
+
+        go_bin = utils_path.find_command('go')
+        cmd = '%s test -v %s -run %s' % (go_bin, self.filename, test_name)
+
+        result = process.run(cmd, ignore_status=True)
+        if result.exit_status != 0:
+            self.fail('Golang Test execution returned a '
+                      'non-0 exit code (%s)' % result)
+
+
+class NotGolangTest(object):
+
+    """
+    Not a golang test (for reporting purposes)
+    """
+
+
+class GolangLoader(loader.TestLoader):
+    """
+    Golang loader class
+    """
+    name = "golang"
+
+    def __init__(self, args, extra_params):
+        super(GolangLoader, self).__init__(args, extra_params)
+
+    def discover(self, url, which_tests=loader.DEFAULT):
+        if url is None:
+            return []
+
+        avocado_suite = []
+        package_paths = []
+        test_files = []
+        subtest = None
+        tests_filter = None
+        go_root = os.environ.get('GOROOT')
+        go_path = os.environ.get('GOPATH')
```

Comment:
I think in case of no `GOPATH` you should include `.`, because that is what I get `unset GOPATH; unsert GOROOT; go test -v . # executes tests from the local directory`

Reply:
Actually the way it works for me is it first looks at `GOROOT` and if not found it looks in `.` and it is not executing both.

So to replicate that you should just add `.` to `package_paths` after the `ROOT/PATH` are inserted.